### PR TITLE
StateMachine

### PR DESCRIPTION
**Pull request by GDYendell**

### CODE REVIEW

File: malcolm/core/controller.py [hunk 1]
```diff
@@ -13,6 +20,16 @@ def __init__(self, block):
         """
         logger_name = "%s.controller" % block.name
         super(Controller, self).__init__(logger_name)
+
+        enum_meta = EnumMeta("State", "State of Block", [])
```
Comment (thomascobb) [inline malcolm/core/controller.py]: Make these in create_attributes

Comment (thomascobb) [inline malcolm/core/controller.py]: And get the State enum labels from the stateMachine state names

File: malcolm/core/runnabledevicestatemachine.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    PRERUN = "PreRun"
+    RUNNING = "Running"
+    POSTRUN = "PostRun"
+    PAUSED = "Paused"
+    REWINDING = "Rewinding"
+    ABORTING = "Aborting"
+    ABORTED = "Aborted"
+
+    def __init__(self, name):
+        super(RunnableDeviceStateMachine, self).__init__(name)
```
Comment (thomascobb) [inline malcolm/core/runnabledevicestatemachine.py]: If we put this at the end, then the superclass can allow transitions from any of our states to Fault or Disabled

File: malcolm/core/statemachine.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        super(StateMachine, self).__init__(logger_name=name)
+
+        self.name = name
+        self.allowed_transitions = OrderedDict()
+        self.busy_states = []
+
+        # Set transitions for standard states
+        self.set_allowed(self.FAULT, [self.RESETTING, self.DISABLED])
+        self.set_allowed(self.DISABLED, self.RESETTING)
+
```
Comment (thomascobb) [inline malcolm/core/statemachine.py]: Add set_allowed for all existing states at this point to FAULT or DISABLED

File: malcolm/core/statemachine.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        Args:
+            initial_state(str): Initial state
+            allowed_states(list(str) / str): States that initial_state can
+            transition to
+        """
+
+        if not isinstance(allowed_states, list):
+            allowed_states = [allowed_states]
+
+        if initial_state in list(self.allowed_transitions.keys()):
```
Comment (thomascobb) [inline malcolm/core/statemachine.py]: using setdefault is probably neater

**Comment by GDYendell** [comment]
Sorry Tom I didn't have time to finish this off as I had to get home.

**Pull request merged**